gapii/cc/pack_encoder.cpp

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright (C) 2017 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+#include "pack_encoder.h"
+#include "core/data/pack/pack.pb.h"
+
+#include <core/cc/stream_writer.h>
+
+#include <google/protobuf/descriptor.pb.h>
+#include <google/protobuf/io/coded_stream.h>
+#include <google/protobuf/message.h>
+
+using ::google::protobuf::Message;
+using ::google::protobuf::Descriptor;
+using ::google::protobuf::DescriptorProto;
+using ::google::protobuf::io::CodedOutputStream;
+using ::google::protobuf::io::StringOutputStream;
+
+namespace {
+
+const uint64_t specialSection = 0;
+const char magic[] = "protopack";
+
+// PackEncoderImpl implements the PackEncoder interface.
+class PackEncoderImpl : public gapii::PackEncoder {
+public:
+    PackEncoderImpl(std::shared_ptr<core::StreamWriter> output);
```

Comment:
I tried moving over to `std::unique_ptr` but the connection is actually shared (it's implemented by `ConnectionStream` that also implements `core::StreamReader`, which is held by other code).
I acknowledge that `unique_ptr` would be preferable here, and I'll see if I can tidy it up later.